tket/tests/TokenSwapping/test_SwapListOptimiser.cpp

```diff
@@ -0,0 +1,495 @@
+
+  EdgesGenerator swaps_generator;
+  vector<Swap> possible_swaps;
+  size_t actual_num_vertices;
+  vector<Swap> raw_swaps;
+
+  void run(
+      RNG& rng, const vector<size_t>& approx_num_vertices,
+      const vector<size_t>& approx_num_edges_percentages,
+      const vector<size_t>& swap_length_percentages,
+      size_t num_tests_per_parameter_list) {
+    for (auto approx_nv : approx_num_vertices) {
+      swaps_generator.approx_num_vertices = approx_nv;
+      for (auto approx_nep : approx_num_edges_percentages) {
+        swaps_generator.approx_num_edges =
+            approx_nv / 2 + (approx_nv * (approx_nv - 1) * approx_nep) / 200;
+        for (size_t num_graphs = 0; num_graphs < 1; ++num_graphs) {
+          possible_swaps = swaps_generator.get_swaps(rng, actual_num_vertices);
+          for (auto slp : swap_length_percentages) {
+            const size_t swap_list_length =
+                1 + (possible_swaps.size() * slp) / 100;
+            for (size_t test_counter = 0;
+                 test_counter < num_tests_per_parameter_list; ++test_counter) {
+              raw_swaps.clear();
+              for (size_t nn = 0; nn < swap_list_length; ++nn) {
+                raw_swaps.push_back(rng.get_element(possible_swaps));
+              }
+              tester.test(raw_swaps);
+            }
+          }
+        }
+      }
+    }
+  }
+};
+}  // namespace
+
+SCENARIO("More realistic swap sequences") {
+  RNG rng;
+  const size_t num_tests_per_parameter_list = 10;
+
+  // How many edges should we aim for, as a rough percentage of
+  // the total number n(n-1)/2 of possibilities?
+  const vector<size_t> approx_num_edges_percentages{5, 10, 20, 30, 40, 80};
+
+  // How long should the swap length be, as a percentage of the
+  // total possible number of swaps?
+  const vector<size_t> swap_length_percentages{50, 100, 200};
+
+  {
+    const vector<size_t> approx_num_vertices{5, 8};
+    ManyTestsRunner runner;
+    runner.run(
+        rng, approx_num_vertices, approx_num_edges_percentages,
+        swap_length_percentages, num_tests_per_parameter_list);
+    CHECK(
+        runner.tester.get_final_result() ==
+        "[ 360 tests; swap counts: 3160  2380  2104  2104  1396  1406 ]");
+  }
+  {
+    const vector<size_t> approx_num_vertices{10, 12, 14};
+    ManyTestsRunner runner;
+    runner.run(
+        rng, approx_num_vertices, approx_num_edges_percentages,
+        swap_length_percentages, num_tests_per_parameter_list);
+    CHECK(
+        runner.tester.get_final_result() ==
+        "[ 540 tests; swap counts: 10370  9048  7580  7580  5180  5216 ]");
+  }
+  {
+    const vector<size_t> approx_num_vertices{30, 35, 40};
+    ManyTestsRunner runner;
+    runner.run(
+        rng, approx_num_vertices, approx_num_edges_percentages,
+        swap_length_percentages, num_tests_per_parameter_list);
+    CHECK(
+        runner.tester.get_final_result() ==
+        "[ 540 tests; swap counts: 38900  37626  30944  30944  24714  "
```

Comment:
Right, so if the code were improved, these tests might start failing.
I think it would be more useful to test something like "the solution must be at least as good as X" (where X is some metric), rather than "the solution must be exactly Y". Then if the code is improved, the tests will still pass.